---
title: "Working with custom transcriptomes using factR2"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{factR2}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
```

```{r setup}
library(factR2)
```

## Introduction 

## Getting started

### Installing factR2

### Materials needed

## Using factR2


### Creating factRObject
### Interacting with factRObject
### Running factR pipeline
### Quantifying additional conservation scores
### Predicting protein domains


**factR2** requires a custom-assembled transcriptome in GTF format. Below is 
a quick-start on using **factR2** using a sample GTF file:


```r
library(factR2)

# get path to sample GTF file
gtf <- system.file("extdata/sc_merged_sample.gtf.gz", package = "factR")

# create factRObject
factR.object <- createfactRObject(gtf, "vM25")

# interacting with factRObject
## print out metadata of various levels
genes(factR.object)  # gene metadata
txs(factR.object)  # transcript metadata
ase(factR.object)  # alternative splicing metadata

## print out transcript metadata for particular genes
txs(factR.object, "Dab2")
txs(factR.object, "Dab2", "Osmr")  # accepts multiple input genes

## View transcript isoforms
plotTranscripts(factR.object, "Dab2")  #from specific genes
plotTranscripts(factR.object, "ENSMUST00000078019.12") # of specific transcripts
### to rescale introns (useful for transcripts with long introns)
plotTranscripts(factR.object, "ENSMUST00000078019.12", rescale_introns=TRUE) 
plotTranscripts(factR.object, "AS00179")  # of specific AS events


# run main factR pipeline
factR.object <- runfactR(factR.object)

# export alternative splicing metadata
exportTable(factR.object, data = "AS")

```