---
title: "Working with custom transcriptomes using factR2"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{factR2}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
```

```{r silent.setup, echo=FALSE}
library(factR2)
library(tidyverse)
library(rtracklayer)
```


```{r setup, eval=FALSE}
library(factR2)
library(tidyverse)
library(rtracklayer)
```

## Introduction 

## Getting started

### Installing factR2

```{r, eval=FALSE}
# install.packages("devtools")
devtools::install_github("f-hamidlab/factR2")
```



### Materials needed

```{r}
gtf.file <-  system.file("extdata/pb_custom.gtf.gz", package = "factR2")
gtf.gr <- import(gtf.file)

head(gtf.gr)
```


## Using factR2

### Creating factRObject

```{r}
listSupportedGenomes()
```
```{r}
fobj <- createfactRObject(gtf.gr, reference = "vM33")
```

### Interacting with factRObject

```{r}
fobj
```

```{r}
listSets(fobj)
```


```{r}
activeSet(fobj) <- "transcript"
```


```{r}
features(fobj)
```

```{r}
features(fobj, "U2af1")
```

```{r}
ase(fobj)
ase(fobj, "U2af1")
```

```{r}
plotTranscripts(fobj, "U2af1")
```


```{r}
plotTranscripts(fobj, "U2af1", rescale_introns = T)
```

### Running factR pipeline

```{r}
features(fobj) %>% 
  group_by(cds) %>% 
  tally()
```




```{r}
fobj <- runfactR(fobj)
```

```{r}
features(fobj) %>% 
  group_by(cds) %>% 
  tally()
```

```{r}
features(fobj) 
```


```{r}
features(fobj) %>% 
  group_by(nmd) %>% 
  tally()
```

```{r}
plotTranscripts(fobj, "U2af1", rescale_introns = T)
```





### Quantifying additional conservation scores

```{r}
fobj <- getAScons(fobj, type = "flanks", padding = 50)
ase(fobj)
```

```{r}
fobj <- getAScons(fobj, type = c("upstream","downstream"), padding = c(50,100))
```

Rethink of what to plot here.

```{r}
ase(fobj) %>% 
  filter(!is.na(ASNMDtype)) %>% 
  ggplot(aes(x=AStype,fill=ASNMDtype)) +
  geom_bar()
```


### Test correlation

```{r}
counts <- system.file("extdata/pb_expression.tsv.gz", package = "factR2")
fobj <- addTxCounts(fobj, counts)
```
```{r}
fobj <- testGeneCorr(fobj, alternative="greater")
ase(fobj) %>% 
  mutate(corr= ifelse(gene.cor.pval < 0.05, "corr", "not-corr")) %>% 
  group_by(ASNMDtype, AStype, corr) %>% 
  tally()
```

```{r}
ase(fobj) %>% 
  filter(gene.cor.pval<0.05, 
         ASNMDtype=="Stimulating", 
         AStype=="CE") %>% arrange(desc(gene.cor.estimate))
```


Plot trend?


### Predicting protein domains

Print AA sequence?

```{r}
plotDomains(fobj, "U2af1")
```




















