<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>Protein-level functions — factR-proteins • factR2</title><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/readable/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><meta property="og:title" content="Protein-level functions — factR-proteins"><meta property="og:description" content='Upon building CDS information on custom transcripts, one can query the translated
sequence of protein-coding transcripts and determine the encoding protein
domains and motifs.
`predictDomains` queries the HMM database (https://www.ebi.ac.uk/Tools/hmmer/search/hmmscan)
for known protein domain families using either the "superfamily" or "pfam"
database. This prediction can be performed globally on all protein coding
transcripts or on specific transcript families (recommended).'><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">
    

    <div class="container template-reference-topic">
      <header><div class="navbar navbar-inverse navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">factR2</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.99.1</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    <span class="fas fa-book-open"></span>
     
    Vignettes
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li>
      <a href="../articles/factR2.html">Using factR2</a>
    </li>
  </ul></li>
<li>
  <a href="../news/index.html">
    <span class="fa fa-newspaper"></span>
     
    News
  </a>
</li>
<li>
  <a href="../reference/index.html">
    <span class="fa fa-code"></span>
     
    Functions
  </a>
</li>
      </ul><ul class="nav navbar-nav navbar-right"><li>
  <a href="https://github.com/fursham-h/factR2" class="external-link">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="https://twitter.com/FurshamH" class="external-link">
    <span class="fab fa-twitter fa-lg"></span>
     
  </a>
</li>
      </ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Protein-level functions</h1>
    
    <div class="hidden name"><code>factR-proteins.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>Upon building CDS information on custom transcripts, one can query the translated
sequence of protein-coding transcripts and determine the encoding protein
domains and motifs.</p>
<p>`predictDomains` queries the HMM database (<a href="https://www.ebi.ac.uk/Tools/hmmer/search/hmmscan" class="external-link">https://www.ebi.ac.uk/Tools/hmmer/search/hmmscan</a>)
for known protein domain families using either the "superfamily" or "pfam"
database. This prediction can be performed globally on all protein coding
transcripts or on specific transcript families (recommended).</p>
    </div>

    <div id="ref-usage">
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="co"># S4 method for factR</span></span>
<span><span class="fu">getAAsequence</span><span class="op">(</span><span class="va">object</span>, verbose <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># S4 method for factR</span></span>
<span><span class="fu">predictDomains</span><span class="op">(</span><span class="va">object</span>, <span class="va">...</span>, database <span class="op">=</span> <span class="st">"superfamily"</span>, ncores <span class="op">=</span> <span class="fl">4</span><span class="op">)</span></span></code></pre></div>
    </div>

    <div id="arguments">
    <h2>Arguments</h2>
    <dl><dt>object</dt>
<dd><p>factRObject</p></dd>


<dt>...</dt>
<dd><p>One or more features to display. Can be the following:</p><ul><li><p>gene_id: ID of gene to plot</p></li>
<li><p>gene_name: Name of gene to plot</p></li>
<li><p>transcript_id: ID of transcript to plot</p></li>
</ul></dd>


<dt>database</dt>
<dd><p>HMM database to query. Can be "superfamily" or "pfam".</p></dd>


<dt>ncores</dt>
<dd><p>Number of cores to run prediction on</p></dd>


<dt>factRObject</dt>
<dd></dd>

</dl></div>
    <div id="value">
    <h2>Value</h2>
    

<p>Updated factRObject.
`getAAsequence` stores an AAStringSet object in the factRObject class.</p>


<p>`predictDomains` stores a dataframe of predicted protein domains in the factRObject.</p>
    </div>

    <div id="ref-examples">
    <h2>Examples</h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="co">## Load sample factRObject and build CDS</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">factRsample</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">factRsample</span> <span class="op">&lt;-</span> <span class="fu"><a href="buildCDS.html">buildCDS</a></span><span class="op">(</span><span class="va">factRsample</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co">## Get peptide sequences</span></span></span>
<span class="r-in"><span><span class="va">factRsample</span> <span class="op">&lt;-</span> <span class="fu">getAAsequence</span><span class="op">(</span><span class="va">factRsample</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co">## Predict domains of gene families</span></span></span>
<span class="r-in"><span><span class="va">factRsample</span> <span class="op">&lt;-</span> <span class="fu">predictDomains</span><span class="op">(</span><span class="va">factRsample</span>, <span class="st">"Osmr"</span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #0000BB;">ℹ </span><span style="color: #0000BB; font-style: italic;">Set `show_more to TRUE to show more info`</span></span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>Skipped 1 non-coding transcripts</span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co">## Predict domains of entire coding transcriptome</span></span></span>
<span class="r-in"><span><span class="co">### This takes some time. Increase `ncores` where necessary</span></span></span>
<span class="r-in"><span><span class="va">factRsample</span> <span class="op">&lt;-</span> <span class="fu">predictDomains</span><span class="op">(</span><span class="va">factRsample</span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #0000BB;">ℹ </span><span style="color: #0000BB; font-style: italic;">Set `show_more to TRUE to show more info`</span></span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>Skipped 70 non-coding transcripts</span>
</code></pre></div>
    </div>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by Fursham Hamid.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

      </footer></div>

  


  

  </body></html>

