<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><meta name="description" content="A factRObject-class contains different types of data at the gene, transcript,
 alternative splicing (AS) and protein domain levels. The functions below are
 designed to display specific contents of a factRObject."><title>Display factR2 object data — factR-meta • factR2</title><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.2.2/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.2.2/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Display factR2 object data — factR-meta"><meta property="og:description" content="A factRObject-class contains different types of data at the gene, transcript,
 alternative splicing (AS) and protein domain levels. The functions below are
 designed to display specific contents of a factRObject."><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-light navbar-expand-lg bg-light"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">factR2</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.99.1</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="nav-item">
  <a class="nav-link" href="../articles/factR2.html">Get started</a>
</li>
<li class="active nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="nav-item">
  <a class="nav-link" href="../news/index.html">Changelog</a>
</li>
      </ul><form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off"></form>

      <ul class="navbar-nav"></ul></div>

    
  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="" class="logo" alt=""><h1>Display factR2 object data</h1>
      
      <div class="d-none name"><code>factR-meta.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>A factRObject-class contains different types of data at the gene, transcript,
 alternative splicing (AS) and protein domain levels. The functions below are
 designed to display specific contents of a factRObject.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="co"># S4 method for factR</span></span>
<span><span class="fu">granges</span><span class="op">(</span><span class="va">object</span>, <span class="va">...</span>, set <span class="op">=</span> <span class="cn">NULL</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># S4 method for factR</span></span>
<span><span class="fu">activeSet</span><span class="op">(</span><span class="va">object</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># S4 method for factR</span></span>
<span><span class="fu">activeSet</span><span class="op">(</span><span class="va">object</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">value</span></span>
<span></span>
<span><span class="co"># S4 method for factR</span></span>
<span><span class="fu">listSets</span><span class="op">(</span><span class="va">object</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># S4 method for factR</span></span>
<span><span class="fu">features</span><span class="op">(</span><span class="va">object</span>, <span class="va">...</span>, set <span class="op">=</span> <span class="cn">NULL</span>, show_more <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># S4 method for factR</span></span>
<span><span class="fu">genes</span><span class="op">(</span><span class="va">object</span>, <span class="va">...</span>, show_more <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># S4 method for factR</span></span>
<span><span class="fu">gns</span><span class="op">(</span><span class="va">object</span>, <span class="va">...</span>, show_more <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># S4 method for factR</span></span>
<span><span class="fu">transcripts</span><span class="op">(</span><span class="va">object</span>, <span class="va">...</span>, show_more <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># S4 method for factR</span></span>
<span><span class="fu">txs</span><span class="op">(</span><span class="va">object</span>, <span class="va">...</span>, show_more <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># S4 method for factR</span></span>
<span><span class="fu">ase</span><span class="op">(</span><span class="va">object</span>, <span class="va">...</span>, show_more <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># S4 method for factR</span></span>
<span><span class="fu">domains</span><span class="op">(</span><span class="va">object</span>, <span class="va">...</span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>
    <dl><dt>object</dt>
<dd><p>factRObject</p></dd>


<dt>...</dt>
<dd><p>One or more features to display. Can be the following:</p><ul><li><p>gene_id: ID of gene to plot</p></li>
<li><p>gene_name: Name of gene to plot</p></li>
<li><p>transcript_id: ID of transcript to plot</p></li>
</ul></dd>


<dt>set</dt>
<dd><p>Set metadata to display. Can be "gene", "transcript" or "AS".</p></dd>


<dt>value</dt>
<dd><p>Character value of one of the following: "gene", "transcript" or "AS"</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    

<ul><li><p>`granges`: GenomicRanges object of selected features</p></li>
<li><p>`activeSet` and `listSets`: Character value/vector</p></li>
<li><p>All other functions: Tibble dataframe containing metadata of selected features</p></li>
</ul></div>

    <div class="section level2">
    <h2 id="ref-examples">Examples<a class="anchor" aria-label="anchor" href="#ref-examples"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="co">### Load sample factRObject</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="st">"factRsample"</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co">## Prints out activeSet</span></span></span>
<span class="r-in"><span><span class="fu">activeSet</span><span class="op">(</span><span class="va">factRsample</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "AS"</span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co">## Change activeSet</span></span></span>
<span class="r-in"><span><span class="fu">activeSet</span><span class="op">(</span><span class="va">factRsample</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="st">"transcript"</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co">## Returns coordinates and metadata of features as a GenomicRanges object</span></span></span>
<span class="r-in"><span><span class="fu">granges</span><span class="op">(</span><span class="va">factRsample</span><span class="op">)</span>   <span class="co"># from activeSet</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> GRanges object with 1149 ranges and 13 metadata columns:</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>          seqnames          ranges strand |               gene_id    source</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>             &lt;Rle&gt;       &lt;IRanges&gt;  &lt;Rle&gt; |           &lt;character&gt;  &lt;factor&gt;</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>      [1]    chr15 3180731-3180944      * |           MSTRG.14523 StringTie</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>      [2]    chr15 3180731-3180944      * |           MSTRG.14523 StringTie</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>      [3]    chr15 3268547-3277274      + | ENSMUSG00000064373.12 StringTie</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>      [4]    chr15 3268547-3268768      + | ENSMUSG00000064373.12 StringTie</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>      [5]    chr15 3274661-3274876      + | ENSMUSG00000064373.12 StringTie</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>      ...      ...             ...    ... .                   ...       ...</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   [1145]    chr15 7305070-7398395      - | ENSMUSG00000042961.13 StringTie</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   [1146]    chr15 7305070-7305219      - | ENSMUSG00000042961.13 StringTie</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   [1147]    chr15 7316399-7316482      - | ENSMUSG00000042961.13 StringTie</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   [1148]    chr15 7318241-7318350      - | ENSMUSG00000042961.13 StringTie</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   [1149]    chr15 7397988-7398395      - | ENSMUSG00000042961.13 StringTie</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>                type   gene_name match_level     score     phase</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>            &lt;factor&gt; &lt;character&gt;   &lt;numeric&gt; &lt;numeric&gt; &lt;integer&gt;</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>      [1] transcript        &lt;NA&gt;           5      1000      &lt;NA&gt;</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>      [2] exon              &lt;NA&gt;           5      1000      &lt;NA&gt;</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>      [3] transcript     Selenop           1      1000      &lt;NA&gt;</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>      [4] exon           Selenop           1      1000      &lt;NA&gt;</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>      [5] exon           Selenop           1      1000      &lt;NA&gt;</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>      ...        ...         ...         ...       ...       ...</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   [1145] transcript      Egflam           1      1000      &lt;NA&gt;</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   [1146] exon            Egflam           1      1000      &lt;NA&gt;</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   [1147] exon            Egflam           1      1000      &lt;NA&gt;</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   [1148] exon            Egflam           1      1000      &lt;NA&gt;</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   [1149] exon            Egflam           1      1000      &lt;NA&gt;</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>                 transcript_id exon_number           ref_gene_id old_gene_id</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>                   &lt;character&gt; &lt;character&gt;           &lt;character&gt; &lt;character&gt;</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>      [1]        MSTRG.14523.1        &lt;NA&gt;                  &lt;NA&gt; MSTRG.14523</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>      [2]        MSTRG.14523.1           1                  &lt;NA&gt; MSTRG.14523</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>      [3] ENSMUST00000160787.8        &lt;NA&gt; ENSMUSG00000064373.12 MSTRG.14525</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>      [4] ENSMUST00000160787.8           1 ENSMUSG00000064373.12 MSTRG.14525</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>      [5] ENSMUST00000160787.8           2 ENSMUSG00000064373.12 MSTRG.14525</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>      ...                  ...         ...                   ...         ...</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   [1145] ENSMUST00000159726.1        &lt;NA&gt; ENSMUSG00000042961.13 MSTRG.14582</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   [1146] ENSMUST00000159726.1           1 ENSMUSG00000042961.13 MSTRG.14582</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   [1147] ENSMUST00000159726.1           2 ENSMUSG00000042961.13 MSTRG.14582</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   [1148] ENSMUST00000159726.1           3 ENSMUSG00000042961.13 MSTRG.14582</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   [1149] ENSMUST00000159726.1           4 ENSMUSG00000042961.13 MSTRG.14582</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>               AStype       AS_id</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>          &lt;character&gt; &lt;character&gt;</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>      [1]        &lt;NA&gt;        &lt;NA&gt;</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>      [2]        &lt;NA&gt;        &lt;NA&gt;</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>      [3]        &lt;NA&gt;        &lt;NA&gt;</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>      [4]        &lt;NA&gt;        &lt;NA&gt;</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>      [5]        &lt;NA&gt;        &lt;NA&gt;</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>      ...         ...         ...</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   [1145]        &lt;NA&gt;        &lt;NA&gt;</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   [1146]        &lt;NA&gt;        &lt;NA&gt;</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   [1147]        &lt;NA&gt;        &lt;NA&gt;</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   [1148]        &lt;NA&gt;        &lt;NA&gt;</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   [1149]        &lt;NA&gt;        &lt;NA&gt;</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   -------</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   seqinfo: 1 sequence from an unspecified genome; no seqlengths</span>
<span class="r-in"><span><span class="fu">granges</span><span class="op">(</span><span class="va">factRsample</span>, <span class="st">"Dab2"</span><span class="op">)</span>   <span class="co"># specific features</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> GRanges object with 142 ranges and 13 metadata columns:</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>         seqnames          ranges strand |               gene_id    source</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>            &lt;Rle&gt;       &lt;IRanges&gt;  &lt;Rle&gt; |           &lt;character&gt;  &lt;factor&gt;</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     [1]    chr15 6299788-6435800      + | ENSMUSG00000022150.16 StringTie</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     [2]    chr15 6299788-6299899      + | ENSMUSG00000022150.16 StringTie</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     [3]    chr15 6416806-6416994      + | ENSMUSG00000022150.16 StringTie</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     [4]    chr15 6418230-6418369      + | ENSMUSG00000022150.16 StringTie</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     [5]    chr15 6419739-6419837      + | ENSMUSG00000022150.16 StringTie</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     ...      ...             ...    ... .                   ...       ...</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   [138]    chr15 6424635-6426667      + | ENSMUSG00000022150.16 StringTie</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   [139]    chr15 6321926-6322138      * | ENSMUSG00000022150.16 StringTie</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   [140]    chr15 6321926-6322138      * | ENSMUSG00000022150.16 StringTie</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   [141]    chr15 6408725-6408926      * | ENSMUSG00000022150.16 StringTie</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   [142]    chr15 6408725-6408926      * | ENSMUSG00000022150.16 StringTie</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>               type   gene_name match_level     score     phase</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>           &lt;factor&gt; &lt;character&gt;   &lt;numeric&gt; &lt;numeric&gt; &lt;integer&gt;</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     [1] transcript        Dab2           1      1000      &lt;NA&gt;</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     [2] exon              Dab2           1      1000      &lt;NA&gt;</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     [3] exon              Dab2           1      1000      &lt;NA&gt;</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     [4] exon              Dab2           1      1000      &lt;NA&gt;</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     [5] exon              Dab2           1      1000      &lt;NA&gt;</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     ...        ...         ...         ...       ...       ...</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   [138] exon              Dab2           1      1000      &lt;NA&gt;</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   [139] transcript        Dab2           4      1000      &lt;NA&gt;</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   [140] exon              Dab2           4      1000      &lt;NA&gt;</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   [141] transcript        Dab2           4      1000      &lt;NA&gt;</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   [142] exon              Dab2           4      1000      &lt;NA&gt;</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>                transcript_id exon_number           ref_gene_id old_gene_id</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>                  &lt;character&gt; &lt;character&gt;           &lt;character&gt; &lt;character&gt;</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     [1] ENSMUST00000159552.1        &lt;NA&gt; ENSMUSG00000022150.16 MSTRG.14562</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     [2] ENSMUST00000159552.1           1 ENSMUSG00000022150.16 MSTRG.14562</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     [3] ENSMUST00000159552.1           2 ENSMUSG00000022150.16 MSTRG.14562</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     [4] ENSMUST00000159552.1           3 ENSMUSG00000022150.16 MSTRG.14562</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     [5] ENSMUST00000159552.1           4 ENSMUSG00000022150.16 MSTRG.14562</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     ...                  ...         ...                   ...         ...</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   [138] ENSMUST00000161558.1           4 ENSMUSG00000022150.16 MSTRG.14562</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   [139]        MSTRG.14564.1        &lt;NA&gt;                  &lt;NA&gt; MSTRG.14564</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   [140]        MSTRG.14564.1           1                  &lt;NA&gt; MSTRG.14564</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   [141]        MSTRG.14566.1        &lt;NA&gt;                  &lt;NA&gt; MSTRG.14566</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   [142]        MSTRG.14566.1           1                  &lt;NA&gt; MSTRG.14566</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>              AStype       AS_id</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>         &lt;character&gt; &lt;character&gt;</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     [1]        &lt;NA&gt;        &lt;NA&gt;</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     [2]        &lt;NA&gt;        &lt;NA&gt;</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     [3]        &lt;NA&gt;        &lt;NA&gt;</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     [4]        &lt;NA&gt;        &lt;NA&gt;</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     [5]        &lt;NA&gt;        &lt;NA&gt;</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     ...         ...         ...</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   [138]        &lt;NA&gt;        &lt;NA&gt;</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   [139]        &lt;NA&gt;        &lt;NA&gt;</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   [140]        &lt;NA&gt;        &lt;NA&gt;</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   [141]        &lt;NA&gt;        &lt;NA&gt;</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   [142]        &lt;NA&gt;        &lt;NA&gt;</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   -------</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   seqinfo: 1 sequence from an unspecified genome; no seqlengths</span>
<span class="r-in"><span><span class="fu">granges</span><span class="op">(</span><span class="va">factRsample</span>, <span class="st">"Dab2"</span>, set <span class="op">=</span> <span class="st">"gene"</span><span class="op">)</span>   <span class="co"># specific features from different Set</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> GRanges object with 1 range and 13 metadata columns:</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>       seqnames          ranges strand |               gene_id   source     type</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>          &lt;Rle&gt;       &lt;IRanges&gt;  &lt;Rle&gt; |           &lt;character&gt; &lt;factor&gt; &lt;factor&gt;</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   [1]    chr15 6299788-6440712      + | ENSMUSG00000022150.16   factR2     gene</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>         gene_name match_level     score     phase transcript_id exon_number</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>       &lt;character&gt;   &lt;numeric&gt; &lt;numeric&gt; &lt;integer&gt;   &lt;character&gt; &lt;character&gt;</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   [1]        Dab2           4        NA      &lt;NA&gt;          &lt;NA&gt;        &lt;NA&gt;</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>       ref_gene_id old_gene_id      AStype       AS_id</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>       &lt;character&gt; &lt;character&gt; &lt;character&gt; &lt;character&gt;</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   [1]        &lt;NA&gt;        &lt;NA&gt;        &lt;NA&gt;        &lt;NA&gt;</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   -------</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   seqinfo: 1 sequence from an unspecified genome; no seqlengths</span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co">## Returns metadata of features</span></span></span>
<span class="r-in"><span><span class="fu">features</span><span class="op">(</span><span class="va">factRsample</span><span class="op">)</span>   <span class="co"># from activeSet</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #0000BB;">ℹ </span><span style="color: #0000BB; font-style: italic;">Set `show_more to TRUE to show more info`</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #949494;"># A tibble: 154 × 8</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    transcript_id         gene_id        gene_name strand width novel cds   nmd  </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>                 <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>          <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>     <span style="color: #949494; font-style: italic;">&lt;fct&gt;</span>  <span style="color: #949494; font-style: italic;">&lt;int&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;"> 1</span> ENSMUST00000022746.12 ENSMUSG000000… Osmr      -       <span style="text-decoration: underline;">5</span>491 no    no    no   </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;"> 2</span> ENSMUST00000175862.7  ENSMUSG000000… Osmr      -       <span style="text-decoration: underline;">2</span>837 no    no    no   </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;"> 3</span> ENSMUST00000176554.1  ENSMUSG000000… Osmr      -        908 no    no    no   </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;"> 4</span> ENSMUST00000176826.1  ENSMUSG000000… Osmr      -       <span style="text-decoration: underline;">4</span>028 no    no    no   </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;"> 5</span> ENSMUST00000177263.1  ENSMUSG000000… Osmr      -       <span style="text-decoration: underline;">6</span>160 no    no    no   </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;"> 6</span> ENSMUST00000177478.1  ENSMUSG000000… Osmr      -       <span style="text-decoration: underline;">1</span>727 no    no    no   </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;"> 7</span> MSTRG.14573.1         ENSMUSG000000… Osmr      *        877 yes   no    no   </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;"> 8</span> ENSMUST00000090461.11 ENSMUSG000000… Fyb       +       <span style="text-decoration: underline;">2</span>678 no    no    no   </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;"> 9</span> ENSMUST00000159604.1  ENSMUSG000000… Fyb       +        771 no    no    no   </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;">10</span> ENSMUST00000159714.1  ENSMUSG000000… Fyb       +       <span style="text-decoration: underline;">2</span>170 no    no    no   </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #949494;"># ℹ 144 more rows</span></span>
<span class="r-in"><span><span class="fu">features</span><span class="op">(</span><span class="va">factRsample</span>, <span class="st">"Dab2"</span><span class="op">)</span>   <span class="co"># specific features</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #0000BB;">ℹ </span><span style="color: #0000BB; font-style: italic;">Set `show_more to TRUE to show more info`</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #949494;"># A tibble: 16 × 8</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    transcript_id         gene_id        gene_name strand width novel cds   nmd  </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>                 <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>          <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>     <span style="color: #949494; font-style: italic;">&lt;fct&gt;</span>  <span style="color: #949494; font-style: italic;">&lt;int&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;"> 1</span> ENSMUST00000078019.12 ENSMUSG000000… Dab2      +       <span style="text-decoration: underline;">3</span>793 no    no    no   </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;"> 2</span> ENSMUST00000080880.11 ENSMUSG000000… Dab2      +       <span style="text-decoration: underline;">4</span>670 no    no    no   </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;"> 3</span> ENSMUST00000110663.8  ENSMUSG000000… Dab2      +       <span style="text-decoration: underline;">3</span>838 no    no    no   </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;"> 4</span> ENSMUST00000110664.8  ENSMUSG000000… Dab2      +       <span style="text-decoration: underline;">4</span>457 no    no    no   </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;"> 5</span> ENSMUST00000159046.1  ENSMUSG000000… Dab2      +        438 no    no    no   </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;"> 6</span> ENSMUST00000159490.7  ENSMUSG000000… Dab2      +        444 no    no    no   </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;"> 7</span> ENSMUST00000159552.1  ENSMUSG000000… Dab2      +        922 no    no    no   </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;"> 8</span> ENSMUST00000160134.7  ENSMUSG000000… Dab2      +       <span style="text-decoration: underline;">1</span>460 no    no    no   </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;"> 9</span> ENSMUST00000161040.7  ENSMUSG000000… Dab2      +       <span style="text-decoration: underline;">1</span>184 no    no    no   </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;">10</span> ENSMUST00000161558.1  ENSMUSG000000… Dab2      +       <span style="text-decoration: underline;">2</span>119 no    no    no   </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;">11</span> ENSMUST00000161812.7  ENSMUSG000000… Dab2      +       <span style="text-decoration: underline;">3</span>774 no    no    no   </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;">12</span> ENSMUST00000162094.2  ENSMUSG000000… Dab2      +        737 no    no    no   </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;">13</span> ENSMUST00000162140.1  ENSMUSG000000… Dab2      +       <span style="text-decoration: underline;">1</span>406 no    no    no   </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;">14</span> ENSMUST00000163082.1  ENSMUSG000000… Dab2      +       <span style="text-decoration: underline;">3</span>839 no    no    no   </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;">15</span> MSTRG.14564.1         ENSMUSG000000… Dab2      *        213 yes   no    no   </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;">16</span> MSTRG.14566.1         ENSMUSG000000… Dab2      *        202 yes   no    no   </span>
<span class="r-in"><span><span class="fu">features</span><span class="op">(</span><span class="va">factRsample</span>, <span class="st">"Dab2"</span>, set <span class="op">=</span> <span class="st">"gene"</span><span class="op">)</span>   <span class="co"># specific features from different Set</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #0000BB;">ℹ </span><span style="color: #0000BB; font-style: italic;">Set `show_more to TRUE to show more info`</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #949494;"># A tibble: 1 × 5</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   gene_id               gene_name strand  width match_level</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>                 <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>     <span style="color: #949494; font-style: italic;">&lt;fct&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;int&gt;</span>       <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;">1</span> ENSMUSG00000022150.16 Dab2      +      <span style="text-decoration: underline;">140</span>925           4</span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co">### This is the same as:</span></span></span>
<span class="r-in"><span><span class="fu">genes</span><span class="op">(</span><span class="va">factRsample</span>, <span class="st">"Dab2"</span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #0000BB;">ℹ </span><span style="color: #0000BB; font-style: italic;">Set `show_more to TRUE to show more info`</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #949494;"># A tibble: 1 × 5</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   gene_id               gene_name strand  width match_level</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>                 <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>     <span style="color: #949494; font-style: italic;">&lt;fct&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;int&gt;</span>       <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;">1</span> ENSMUSG00000022150.16 Dab2      +      <span style="text-decoration: underline;">140</span>925           4</span>
<span class="r-in"><span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co">## To return protein-coding domains, the protein-coding domains need to be predicted first:</span></span></span>
<span class="r-in"><span><span class="va">factRsample</span> <span class="op">&lt;-</span> <span class="fu"><a href="buildCDS.html">buildCDS</a></span><span class="op">(</span><span class="va">factRsample</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">factRsample</span> <span class="op">&lt;-</span> <span class="fu">getAAsequence</span><span class="op">(</span><span class="va">factRsample</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">factRsample</span> <span class="op">&lt;-</span> <span class="fu">predictDomains</span><span class="op">(</span><span class="va">factRsample</span>, <span class="st">"Dab2"</span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #0000BB;">ℹ </span><span style="color: #0000BB; font-style: italic;">Set `show_more to TRUE to show more info`</span></span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>Skipped 6 non-coding transcripts</span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co">## Then, the domains of the selected gene can be printed as such:</span></span></span>
<span class="r-in"><span><span class="fu">domains</span><span class="op">(</span><span class="va">factRsample</span>, <span class="st">"Dab2"</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #949494;"># A tibble: 8 × 6</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   type   description    eval    begin   end transcript_id        </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>  <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>          <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>                </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;">1</span> DOMAIN PH domain-like 2.7e-42    33   180 ENSMUST00000078019.12</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;">2</span> DOMAIN PH domain-like 4.9e-42    33   180 ENSMUST00000080880.11</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;">3</span> DOMAIN PH domain-like 2.7e-42    33   180 ENSMUST00000110663.8 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;">4</span> DOMAIN PH domain-like 4.7e-42    33   180 ENSMUST00000110664.8 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;">5</span> DOMAIN PH domain-like 3.9e-43    33   180 ENSMUST00000160134.7 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;">6</span> DOMAIN PH domain-like 3.8e-43    33   180 ENSMUST00000161040.7 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;">7</span> DOMAIN PH domain-like 2.5e-42    33   180 ENSMUST00000161812.7 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;">8</span> DOMAIN PH domain-like 2.6e-21    35   110 ENSMUST00000162094.2 </span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co">## All outputs can be assigned to a variable and manipulated further using other functions:</span></span></span>
<span class="r-in"><span><span class="fu">ase</span><span class="op">(</span><span class="va">factRsample</span><span class="op">)</span> <span class="op">%&gt;%</span> <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">AStype</span> <span class="op">==</span> <span class="st">"CE"</span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #0000BB;">ℹ </span><span style="color: #0000BB; font-style: italic;">Set `show_more to TRUE to show more info`</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #949494;"># A tibble: 37 × 7</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    AS_id   gene_id               gene_name coord             AStype strand width</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>                 <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>     <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>             <span style="color: #949494; font-style: italic;">&lt;fct&gt;</span>  <span style="color: #949494; font-style: italic;">&lt;fct&gt;</span>  <span style="color: #949494; font-style: italic;">&lt;int&gt;</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;"> 1</span> AS00005 ENSMUSG00000064373.12 Selenop   chr15:3272755-32… CE     +         51</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;"> 2</span> AS00029 ENSMUSG00000022186.14 Oxct1     chr15:4059936-40… CE     +         61</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;"> 3</span> AS00030 ENSMUSG00000022186.14 Oxct1     chr15:4091168-40… CE     +        108</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;"> 4</span> AS00031 ENSMUSG00000022186.14 Oxct1     chr15:4092411-40… CE     +        115</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;"> 5</span> AS00033 ENSMUSG00000022186.14 Oxct1     chr15:4094015-40… CE     +         95</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;"> 6</span> AS00034 ENSMUSG00000022186.14 Oxct1     chr15:4096472-40… CE     +         49</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;"> 7</span> AS00035 ENSMUSG00000022186.14 Oxct1     chr15:4101127-41… CE     +         73</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;"> 8</span> AS00036 ENSMUSG00000022186.14 Oxct1     chr15:4101804-41… CE     +         76</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;"> 9</span> AS00037 ENSMUSG00000022186.14 Oxct1     chr15:4128834-41… CE     +         90</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;">10</span> AS00039 ENSMUSG00000022186.14 Oxct1     chr15:4142792-41… CE     +         81</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #949494;"># ℹ 27 more rows</span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span></span></span>
</code></pre></div>
    </div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p></p><p>Developed by Fursham Hamid.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

    </footer></div>

  

  

  </body></html>

