<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><meta name="description" content='Upon building CDS information on custom transcripts, one can query the translated
sequence of protein-coding transcripts and determine the encoding protein
domains and motifs.
`predictDomains` queries the HMM database (https://www.ebi.ac.uk/Tools/hmmer/search/hmmscan)
for known protein domain families using either the "superfamily" or "pfam"
database. This prediction can be performed globally on all protein coding
transcripts or on specific transcript families (recommended).'><title>Protein-level functions — factR-proteins • factR2</title><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.2.2/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.2.2/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Protein-level functions — factR-proteins"><meta property="og:description" content='Upon building CDS information on custom transcripts, one can query the translated
sequence of protein-coding transcripts and determine the encoding protein
domains and motifs.
`predictDomains` queries the HMM database (https://www.ebi.ac.uk/Tools/hmmer/search/hmmscan)
for known protein domain families using either the "superfamily" or "pfam"
database. This prediction can be performed globally on all protein coding
transcripts or on specific transcript families (recommended).'><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-light navbar-expand-lg bg-light"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">factR2</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.99.1</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="nav-item">
  <a class="nav-link" href="../articles/factR2.html">Get started</a>
</li>
<li class="active nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="nav-item">
  <a class="nav-link" href="../news/index.html">Changelog</a>
</li>
      </ul><form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off"></form>

      <ul class="navbar-nav"></ul></div>

    
  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="" class="logo" alt=""><h1>Protein-level functions</h1>
      
      <div class="d-none name"><code>factR-proteins.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>Upon building CDS information on custom transcripts, one can query the translated
sequence of protein-coding transcripts and determine the encoding protein
domains and motifs.</p>
<p>`predictDomains` queries the HMM database (<a href="https://www.ebi.ac.uk/Tools/hmmer/search/hmmscan" class="external-link">https://www.ebi.ac.uk/Tools/hmmer/search/hmmscan</a>)
for known protein domain families using either the "superfamily" or "pfam"
database. This prediction can be performed globally on all protein coding
transcripts or on specific transcript families (recommended).</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="co"># S4 method for factR</span></span>
<span><span class="fu">getAAsequence</span><span class="op">(</span><span class="va">object</span>, verbose <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># S4 method for factR</span></span>
<span><span class="fu">predictDomains</span><span class="op">(</span><span class="va">object</span>, <span class="va">...</span>, database <span class="op">=</span> <span class="st">"superfamily"</span>, ncores <span class="op">=</span> <span class="fl">4</span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>
    <dl><dt>object</dt>
<dd><p>factRObject</p></dd>


<dt>...</dt>
<dd><p>One or more features to display. Can be the following:</p><ul><li><p>gene_id: ID of gene to plot</p></li>
<li><p>gene_name: Name of gene to plot</p></li>
<li><p>transcript_id: ID of transcript to plot</p></li>
</ul></dd>


<dt>database</dt>
<dd><p>HMM database to query. Can be "superfamily" or "pfam".</p></dd>


<dt>ncores</dt>
<dd><p>Number of cores to run prediction on</p></dd>


<dt>factRObject</dt>
<dd></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    

<p>Updated factRObject.
`getAAsequence` stores an AAStringSet object in the factRObject class.</p>


<p>`predictDomains` stores a dataframe of predicted protein domains in the factRObject.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-examples">Examples<a class="anchor" aria-label="anchor" href="#ref-examples"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="co">## Load sample factRObject and build CDS</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">factRsample</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">factRsample</span> <span class="op">&lt;-</span> <span class="fu"><a href="buildCDS.html">buildCDS</a></span><span class="op">(</span><span class="va">factRsample</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co">## Get peptide sequences</span></span></span>
<span class="r-in"><span><span class="va">factRsample</span> <span class="op">&lt;-</span> <span class="fu">getAAsequence</span><span class="op">(</span><span class="va">factRsample</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co">## Predict domains of gene families</span></span></span>
<span class="r-in"><span><span class="va">factRsample</span> <span class="op">&lt;-</span> <span class="fu">predictDomains</span><span class="op">(</span><span class="va">factRsample</span>, <span class="st">"Osmr"</span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #0000BB;">ℹ </span><span style="color: #0000BB; font-style: italic;">Set `show_more to TRUE to show more info`</span></span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>Skipped 1 non-coding transcripts</span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co">## Predict domains of entire coding transcriptome</span></span></span>
<span class="r-in"><span><span class="co">### This takes some time. Increase `ncores` where necessary</span></span></span>
<span class="r-in"><span><span class="va">factRsample</span> <span class="op">&lt;-</span> <span class="fu">predictDomains</span><span class="op">(</span><span class="va">factRsample</span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #0000BB;">ℹ </span><span style="color: #0000BB; font-style: italic;">Set `show_more to TRUE to show more info`</span></span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>Skipped 70 non-coding transcripts</span>
</code></pre></div>
    </div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p></p><p>Developed by Fursham Hamid.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

    </footer></div>

  

  

  </body></html>

