<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><meta name="description" content="A container for storing and processing custom transcriptome."><title>factRObject Class — factRObject-class • factR2</title><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.2.2/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.2.2/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="factRObject Class — factRObject-class"><meta property="og:description" content="A container for storing and processing custom transcriptome."><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-light navbar-expand-lg bg-light"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">factR2</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.99.1</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="nav-item">
  <a class="nav-link" href="../articles/factR2.html">Get started</a>
</li>
<li class="active nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="nav-item">
  <a class="nav-link" href="../news/index.html">Changelog</a>
</li>
      </ul><form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off"></form>

      <ul class="navbar-nav"></ul></div>

    
  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="" class="logo" alt=""><h1>factRObject Class</h1>
      
      <div class="d-none name"><code>factRObject-class.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>A container for storing and processing custom transcriptome.</p>
    </div>


    <div class="section level2">
    <h2 id="details">Details<a class="anchor" aria-label="anchor" href="#details"></a></h2>
    <p>The factRObject is a representation of custom-assembled transcriptome data.
Coordinate and metadata information from an input transcriptome (GTF) is
parsed and grouped into 3 data Sets which corresponds to the levels of gene (genes),
transcript (txs) and alternative splicing (ase). With the aid of a compatible
reference genome and expression data, functional and
comparative analyses can be performed to better understand the complexity
of a transcriptome.</p>
    </div>
    <div class="section level2">
    <h2 id="constructor">Constructor<a class="anchor" aria-label="anchor" href="#constructor"></a></h2>
    

<dl><dt>factRObject can be easily constructed as such:</dt>
<dd><p><code><code><a href="createfactRObject.html">createfactRObject</a></code>(gtf, reference)</code></p></dd>


</dl></div>
    <div class="section level2">
    <h2 id="accessors">Accessors<a class="anchor" aria-label="anchor" href="#accessors"></a></h2>
    

<p>Interact with a factRObject (x) the following ways:</p><dl><dt></dt>
<dd><p><code>summary(x): Prints preview of factRObject</code></p></dd>

   <dt></dt>
<dd><p><code>head(x): Prints the first 6 metadata entries of the active Set</code></p></dd>

   <dt></dt>
<dd><p><code>tail(x): Prints the last 6 metadata entries of the active Set</code></p></dd>

   <dt></dt>
<dd><p><code>dim(x): Prints the dimension of the active set (features x samples)</code></p></dd>

   <dt></dt>
<dd><p><code>nrow(x): Returns the number of features of the active set</code></p></dd>

   <dt></dt>
<dd><p><code>ncol(x): Returns the number of samples</code></p></dd>


</dl><p>A factRObject consists of 3 `Sets` of data: gene, transcript and AS. At any
one time, one of these `Sets` will be made the active Set and some factR2
functions will be called only on this active Set. Below are some ways to
check and change the active Set:</p><dl><dt></dt>
<dd><p><code><code>listSets</code>(x): Lists Sets in object</code></p></dd>

   <dt></dt>
<dd><p><code><code>activeSet</code>(x): Returns the active Set of object</code></p></dd>

   <dt></dt>
<dd><p><code><code>activeSet</code>(x)&lt;-: Change the active Set</code></p></dd>


</dl><p>Each `Set` of an factRObject hold rich information of the features that it
contains. Below are some convenient wrappers to display the metadata in
each `Set`.</p><dl><dt></dt>
<dd><p><code><code>features</code>(x): Displays metadata of active Set</code></p></dd>

   <dt></dt>
<dd><p><code><code>genes</code>(x) or <code>gns</code>(x): Displays gene metadata</code></p></dd>

   <dt></dt>
<dd><p><code><code>transcripts</code>(x) or <code>txs</code>(x): Displays transcript metadata</code></p></dd>

   <dt></dt>
<dd><p><code><code>ase</code>(x): Displays alternative splicing events metadata</code></p></dd>


</dl><p>factR2 also contain functions to plot exon-level and domain-level
structures as an interactive plot:</p><dl><dt></dt>
<dd><p><code><code>plotTranscripts</code>(x, "gene of interest"): Plots transcripts</code></p></dd>

   <dt></dt>
<dd><p><code><code>plotDomains</code>(x, "gene of interest")): Plots domains</code></p></dd>


</dl><p>If transcript-level counts data is provided, additional accessors are available:</p><dl><dt></dt>
<dd><p><code><code>counts</code>(x): View/extract counts data</code></p></dd>

   <dt></dt>
<dd><p><code><code>samples</code>(x)): Prints sample metadata</code></p></dd>

   <dt></dt>
<dd><p><code><code>rowNames</code>(x)): Print feature names of current Set</code></p></dd>

   <dt></dt>
<dd><p><code><code>colNames</code>(x)): Prints sample names</code></p></dd>

   <dt></dt>
<dd><p><code><code>colNames</code>(x))&lt;-: Set new sample names</code></p></dd>

   <dt></dt>
<dd><p><code><code>ident</code>(x)): Prints default sample identity</code></p></dd>

   <dt></dt>
<dd><p><code><code>ident</code>(x))&lt;-: Set new sample identity</code></p></dd>


</dl></div>
    <div class="section level2">
    <h2 id="see-also">See also<a class="anchor" aria-label="anchor" href="#see-also"></a></h2>
    <div class="dont-index"><p><code><a href="factR-meta.html">factR-meta</a></code> <code><a href="factR-exp-meta.html">factR-exp-meta</a></code></p></div>
    </div>

  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p></p><p>Developed by Fursham Hamid.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

    </footer></div>

  

  

  </body></html>

