<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><meta name="description" content="This wrapper function will perform the core factR workflow which
includes:
1:  Building CDS structures
2:  Predicting NMD-sensitive transcripts
3:  Translating open-reading frames
4:  Quantify sequence conservation of alternative exons

"><title>Run factR workflow — runfactR • factR2</title><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.2.2/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.2.2/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Run factR workflow — runfactR"><meta property="og:description" content="This wrapper function will perform the core factR workflow which
includes:
1:  Building CDS structures
2:  Predicting NMD-sensitive transcripts
3:  Translating open-reading frames
4:  Quantify sequence conservation of alternative exons

"><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-light navbar-expand-lg bg-light"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">factR2</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.99.1</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="nav-item">
  <a class="nav-link" href="../articles/factR2.html">Get started</a>
</li>
<li class="active nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="nav-item">
  <a class="nav-link" href="../news/index.html">Changelog</a>
</li>
      </ul><form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off"></form>

      <ul class="navbar-nav"></ul></div>

    
  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="" class="logo" alt=""><h1>Run factR workflow</h1>
      
      <div class="d-none name"><code>factR-run.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>This wrapper function will perform the core factR workflow which
includes:</p><ul><li><p>1:  Building CDS structures</p></li>
<li><p>2:  Predicting NMD-sensitive transcripts</p></li>
<li><p>3:  Translating open-reading frames</p></li>
<li><p>4:  Quantify sequence conservation of alternative exons</p></li>
</ul></div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">runfactR</span><span class="op">(</span></span>
<span>  <span class="va">object</span>,</span>
<span>  NMD_threshold <span class="op">=</span> <span class="fl">50</span>,</span>
<span>  cons_db <span class="op">=</span> <span class="st">"phastCons"</span>,</span>
<span>  cons_type <span class="op">=</span> <span class="st">"exon"</span>,</span>
<span>  cons_padding <span class="op">=</span> <span class="fl">0</span>,</span>
<span>  verbose <span class="op">=</span> <span class="cn">TRUE</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>
    <dl><dt>object</dt>
<dd><p>factR object</p></dd>


<dt>NMD_threshold</dt>
<dd><p>Minimum distance between PTC and downstream exon-exon junction to trigger NMD (Default: 50)</p></dd>


<dt>cons_db</dt>
<dd><p>Database to extract sequence conservation. Can be "phastCons" (Default) or "phylop"</p></dd>


<dt>cons_type</dt>
<dd><p>Feature to quantify conservation. Can be one of the following:</p><ul><li><p>"exon" : Sequence conservation of the entire exon</p></li>
<li><p>"flanks" : Conservation of sequences flanking exons</p></li>
<li><p>"upstream" : Conservation of sequences upstream of exons</p></li>
<li><p>"downstream" : Conservation of sequences downstream of exons</p></li>
</ul></dd>


<dt>cons_padding</dt>
<dd><p>Additional width to pad the sequence by. For cons_type "exons" and "flanks", paddings will be added on both sides.</p></dd>


<dt>verbose</dt>
<dd><p>Whether to show pipeline messages (Default: TRUE)</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    

<p>Update factRObject with additional data from runfactR. See links below
for detailed description of each function in this wrapper.</p>
    </div>
    <div class="section level2">
    <h2 id="see-also">See also<a class="anchor" aria-label="anchor" href="#see-also"></a></h2>
    <div class="dont-index"><p><code><a href="buildCDS.html">buildCDS</a></code>, <code>predictNMD</code>, <code>testASNMDevents</code>, <code>getAScons</code>, <code>getAAsequence</code></p></div>
    </div>

    <div class="section level2">
    <h2 id="ref-examples">Examples<a class="anchor" aria-label="anchor" href="#ref-examples"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="co">## Load sample factRObject</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">factRsample</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co">## Run factR pipeline</span></span></span>
<span class="r-in"><span><span class="va">factRsample</span> <span class="op">&lt;-</span> <span class="fu">runfactR</span><span class="op">(</span><span class="va">factRsample</span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #00BB00; font-weight: bold;">🡆  Building CDS information</span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="font-style: italic;">    Searching for reference mRNAs in query</span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="font-style: italic;">    22 reference mRNAs found and its CDS were assigned</span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="font-style: italic;">    Building database of annotated ATG codons</span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="font-style: italic;">    Selecting best ATG start codon for remaining transcripts and determining open-reading frame</span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="font-style: italic;">    61 new CDSs constructed</span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="font-style: italic;">    Summary: Out of 155 transcripts in `gtf`,</span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="font-style: italic;">    83 transcript CDSs were built</span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #0000BB;">    ℹ </span><span style="color: #0000BB; font-style: italic;">Updating transcript feature data</span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #00BB00; font-weight: bold;">🡆  Running transcript-level NMD prediction</span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="font-style: italic;">    Predicting NMD sensitivities for 83 mRNAs</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   |                                                  | 0 % ~calculating    |+                                                 | 1 % ~00s            |++                                                | 2 % ~00s            |++                                                | 4 % ~00s            |+++                                               | 5 % ~00s            |++++                                              | 6 % ~00s            |++++                                              | 7 % ~00s            |+++++                                             | 8 % ~00s            |+++++                                             | 10% ~00s            |++++++                                            | 11% ~00s            |+++++++                                           | 12% ~00s            |+++++++                                           | 13% ~00s            |++++++++                                          | 14% ~00s            |++++++++                                          | 16% ~00s            |+++++++++                                         | 17% ~00s            |++++++++++                                        | 18% ~00s            |++++++++++                                        | 19% ~00s            |+++++++++++                                       | 20% ~00s            |+++++++++++                                       | 22% ~00s            |++++++++++++                                      | 23% ~00s            |+++++++++++++                                     | 24% ~00s            |+++++++++++++                                     | 25% ~00s            |++++++++++++++                                    | 27% ~00s            |++++++++++++++                                    | 28% ~00s            |+++++++++++++++                                   | 29% ~00s            |++++++++++++++++                                  | 30% ~00s            |++++++++++++++++                                  | 31% ~00s            |+++++++++++++++++                                 | 33% ~00s            |+++++++++++++++++                                 | 34% ~00s            |++++++++++++++++++                                | 35% ~00s            |+++++++++++++++++++                               | 36% ~00s            |+++++++++++++++++++                               | 37% ~00s            |++++++++++++++++++++                              | 39% ~00s            |++++++++++++++++++++                              | 40% ~00s            |+++++++++++++++++++++                             | 41% ~00s            |++++++++++++++++++++++                            | 42% ~00s            |++++++++++++++++++++++                            | 43% ~00s            |+++++++++++++++++++++++                           | 45% ~00s            |+++++++++++++++++++++++                           | 46% ~00s            |++++++++++++++++++++++++                          | 47% ~00s            |+++++++++++++++++++++++++                         | 48% ~00s            |+++++++++++++++++++++++++                         | 49% ~00s            |++++++++++++++++++++++++++                        | 51% ~00s            |++++++++++++++++++++++++++                        | 52% ~00s            |+++++++++++++++++++++++++++                       | 53% ~00s            |++++++++++++++++++++++++++++                      | 54% ~00s            |++++++++++++++++++++++++++++                      | 55% ~00s            |+++++++++++++++++++++++++++++                     | 57% ~00s            |+++++++++++++++++++++++++++++                     | 58% ~00s            |++++++++++++++++++++++++++++++                    | 59% ~00s            |+++++++++++++++++++++++++++++++                   | 60% ~00s            |+++++++++++++++++++++++++++++++                   | 61% ~00s            |++++++++++++++++++++++++++++++++                  | 63% ~00s            |++++++++++++++++++++++++++++++++                  | 64% ~00s            |+++++++++++++++++++++++++++++++++                 | 65% ~00s            |++++++++++++++++++++++++++++++++++                | 66% ~00s            |++++++++++++++++++++++++++++++++++                | 67% ~00s            |+++++++++++++++++++++++++++++++++++               | 69% ~00s            |+++++++++++++++++++++++++++++++++++               | 70% ~00s            |++++++++++++++++++++++++++++++++++++              | 71% ~00s            |+++++++++++++++++++++++++++++++++++++             | 72% ~00s            |+++++++++++++++++++++++++++++++++++++             | 73% ~00s            |++++++++++++++++++++++++++++++++++++++            | 75% ~00s            |++++++++++++++++++++++++++++++++++++++            | 76% ~00s            |+++++++++++++++++++++++++++++++++++++++           | 77% ~00s            |++++++++++++++++++++++++++++++++++++++++          | 78% ~00s            |++++++++++++++++++++++++++++++++++++++++          | 80% ~00s            |+++++++++++++++++++++++++++++++++++++++++         | 81% ~00s            |+++++++++++++++++++++++++++++++++++++++++         | 82% ~00s            |++++++++++++++++++++++++++++++++++++++++++        | 83% ~00s            |+++++++++++++++++++++++++++++++++++++++++++       | 84% ~00s            |+++++++++++++++++++++++++++++++++++++++++++       | 86% ~00s            |++++++++++++++++++++++++++++++++++++++++++++      | 87% ~00s            |++++++++++++++++++++++++++++++++++++++++++++      | 88% ~00s            |+++++++++++++++++++++++++++++++++++++++++++++     | 89% ~00s            |++++++++++++++++++++++++++++++++++++++++++++++    | 90% ~00s            |++++++++++++++++++++++++++++++++++++++++++++++    | 92% ~00s            |+++++++++++++++++++++++++++++++++++++++++++++++   | 93% ~00s            |+++++++++++++++++++++++++++++++++++++++++++++++   | 94% ~00s            |++++++++++++++++++++++++++++++++++++++++++++++++  | 95% ~00s            |+++++++++++++++++++++++++++++++++++++++++++++++++ | 96% ~00s            |+++++++++++++++++++++++++++++++++++++++++++++++++ | 98% ~00s            |++++++++++++++++++++++++++++++++++++++++++++++++++| 99% ~00s            |++++++++++++++++++++++++++++++++++++++++++++++++++| 100% elapsed=00s  </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #0000BB;">    ℹ </span><span style="color: #0000BB; font-style: italic;">Updating transcript feature data</span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #00BB00; font-weight: bold;">🡆  Translating amino acid sequences</span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #BB0000;">    ❗ </span><span style="color: #00BBBB; font-style: italic;">1 CDSs do not begin with ATG</span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #BB0000;">    ❗ </span><span style="color: #00BBBB; font-style: italic;">2 CDSs contain internal stop codon. Truncating CDS sequence to retain ORF</span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #00BB00; font-weight: bold;">🡆  Running AS-NMD testing</span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #0000BB;">    ℹ </span><span style="color: #0000BB; font-style: italic;">Getting best reference for AS-NMD testing</span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #0000BB;">    ℹ </span><span style="color: #0000BB; font-style: italic;">Testing AS-NMD exons</span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #0000BB;">    ℹ </span><span style="color: #0000BB; font-style: italic;">Updating AS feature data</span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #00BB00; font-weight: bold;">🡆  Quantifiying conservation scores</span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #0000BB;">    ℹ </span><span style="color: #0000BB; font-style: italic;">Retrieving phastCons60way.UCSC.mm10 scores</span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #0000BB;">    ℹ </span><span style="color: #0000BB; font-style: italic;">Quantifying `exon` conservation scores with 0 padding</span></span>
</code></pre></div>
    </div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p></p><p>Developed by Fursham Hamid.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

    </footer></div>

  

  

  </body></html>

